### PR TITLE
Migrate parfor to SPIRVKernelDispatcher

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
What was done:
- Compile parfor kernels using `SpirvDispatcher`.
- Add internal option to compile kernel with Prange support. This mode inserts prange body instead of `__seninel__ = 0`. Ideally it  must be replace to some internal dedicated function call like `numba_dpex.core.insert_parfor_body()`.
- Remove kernel allocation aside of dpjit compilation. Now parfor kernel specific SPIRV is a part of dpjit function inserted as a string constant. It unblocks dpjit function caching on disk

Checklist:
- [x] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/core/parfors/kernel_builder.py [hunk 1]
```diff
@@ -367,120 +249,33 @@ def create_kernel_for_parfor(
...
-            print("No aliases found so adding noalias flag.")
-        flags.noalias = True
-
-    remove_dead(kernel_ir.blocks, kernel_ir.arg_names, kernel_ir, typemap)
-
-    if config.DEBUG_ARRAY_OPT:
-        print("kernel_ir after remove dead")
-        kernel_ir.dump()
-
-    # The first argument to a range kernel is a kernel_api.Item object. The
```
Comment (diptorupd) [inline numba_dpex/core/parfors/kernel_builder.py:429]: Maybe keep the code comment around.

File: numba_dpex/core/parfors/kernel_parfor_pass.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+from typing import NamedTuple, Union
```
Comment (diptorupd) [inline numba_dpex/core/parfors/kernel_parfor_pass.py:1]: license header is missing

Comment (diptorupd) [reply numba_dpex/core/parfors/kernel_parfor_pass.py:1]: I also think `parfor_kernel_pass` or `parfor_to_kernel_pass` is a better name as a parfor is being converted to a kernel.

Comment (ZzEeKkAa) [reply numba_dpex/core/parfors/kernel_parfor_pass.py:1]: I guess `parfor_inject_kernel_pass` is a better name

File: numba_dpex/core/pipelines/kernel_compiler.py [hunk 3]
```diff
@@ -56,6 +58,16 @@ def define_nopython_pipeline(state, name="dpex_kernel_nopython"):
         ndpb = DefaultPassBuilder
         pm = PassManager(name)
         untyped_passes = ndpb.define_untyped_pipeline(state)
+
+        for i, _pass in enumerate(untyped_passes.passes):
```
Comment (ZzEeKkAa) [inline numba_dpex/core/pipelines/kernel_compiler.py:62]: TODO:
pm.add_pass_after(MlirBackend, NopythonTypeInference)

Comment (ZzEeKkAa) [reply numba_dpex/core/pipelines/kernel_compiler.py:62]: Move it into separate pass.

**Review by diptorupd — APPROVED** [review]
There are few cleanups such as introducing a new pipeline fr parfor kernel compilation that should be done, can be handled in later PR.

Overall nice set of changes. The parfor kernel generator has become much cleaner after these changes.

**Pull request merged**